### ISSUE TITLE
On Scales & Drivers

### ISSUE THREAD

**Issue by gohanman** [CORE-POS/IS4C#682]
(This is partially me thinking out loud)

Consistent scale behavior has been a frustratingly intermittent problem. A non-trivial part of this has to do with the scale's communication protocol.

### States
At any given time the scale can be in 1 of 6 states:
* State 0 is "not ready". I've never really run into this. It might occur briefly on initial boot or in the case of hardware failure where it literally cannot weigh.
* State 1 is "in motion". This is a very common state.
* State 2 is "over max weight". This is rare but straightforward.
* State 3 is "stable zero weight". This is a very common state.
* State 4 is "stable non-zero weight". This is a very common state.
* State 5 is "under zero weight". Could be as simple as needing to press the zero key but could also be a leveling or calibration problem.

### Commands
The scale [basically] supports two different commands.
* `S11` is "get next stable non-zero weight"
* `S14` is "get current status"

In either case the scale will send one response to the command.

### Problems
Both `SPH_Magellan_Scale` and `SPH_Magellan_Classic` use a mix of the two commands and both approaches have drawbacks. `SPH_Magellan_Scale` makes heavier use of `S14` where as `SPH_Magellan_Classic` makes heavier use of `S11`.

Using `S14` is just polling. The scale and driver are exchanging status constantly. I suspect this leads to slow downs - especially if the driver is handing duplicate status messages back up to POS. But pausing or sleeping between status checks just increases slow downs.

Using `S11` is vastly more efficient but the fact that it only returns stable *non-zero* weights is a significant drawback. If the scale is in motion when you send an `S11` and it settles at zero the scale's going to appear to be stuck. I think most situations where `WAKEUP` fixes a stuck scale are some variation of this problem. All `WAKEUP` really does is send an `S14`.

### The Beep of Death
My theory continues to be that this is a concurrency issue. My top guesses would be:
* Sending 2+ `S14` commands without waiting for an acknowledgement
* Interleaving 2 commands in a way that simply confuses the scale
* Interleaving 2 commands to create a "Disable scale" command. I cannot find any documentation of what that command might be, but `S13` can report the scale status as enabled or disabled and there are commands to enable or disable the scanner so there may be equivalents for the scale.

Under all theories I think a constant stream of `S14`s is most likely to trigger problems. I am curious if `S10`, hard reset, will awaken a beep-of-death'd scale but without a reliable way to trigger that state it's difficult to test.

### Solutions
I have a couple different ideas. First would be to rely *even more* heavily on `S14`. Concurrency would need to be addressed through locking and the driver would have to keep track of recent states to avoid flooding POS with scale status messages. But `WAKEUP` would theoretically be obsolete if the response to every status is another request-for-status.

Alternately, we could respond to *all* scale statuses with `S11` and track state separately. An independent timer could track whether a status check is needed. If there hasn't been a message from the scale in X milliseconds, cancel the current `S11` and send an `S14` - but then respond to that status with another `S11`. This would involve a lot less [mostly redundant] data going back and forth between the driver and the scale. The question is whether there's a workable value for X in milliseconds that won't feel laggy. Non-zero weight reads should be instantaneous but dropping back to stable zero would take at least X milliseconds.

Other ideas are of course welcome. I don't have a strong sense of what the better option is. Eliminating `WAKEUP` one way or another does seem feasible though. The driver ought to be able to notice a lack of info from the scale faster than an end user.

**Comment by gohanman** [comment]
Attempt at the `S14` only solution: 8856ca2137f97c2bf10e68155b40356818df4bfa

Scale state is tracked with every status and messages are only passed to POS when the state changes. Locking should prevent Beeps, `WAKEUP`s, and normal polling from overlapping.

I have not actually tried this yet.

**Comment by gohanman** [comment]
Follow-up after trying it: weight response are noticeably faster. Even with the the attempts to limit messages to POS to once per state change it sends a *lot* more messages. The problem that's immediately apparent is under the current message passing system there's no way to guarantee POS processes the messages in the same order the driver produces them.

I don't think going by file create/modified time is helpful. It could easily be more than one message/second and that introduces cross-platform headaches in what level of time precision the file system maintains. I think prefixing the file names with an integer and then sorting on the receiving end would work if it can somehow detect lulls and pre-emptively reset the counter to zero when the message directory is empty. Otherwise they're going to be random weirdness every INTMAX messages.

There could also be hardware specific I/O issues where this approach is faster with SSD but slower with spinny disk.

**Comment by flathat** [comment]
There used to be discussion of using RAM disk; was it for this?

**Comment by gohanman** [comment]
I think it was in the context of the old ssd.c driver although the same general principle would apply that a RAM disk would be faster than a spinning disk. But the difference between a RAM disk and solid state drive would be pretty marginal. 

I can't see much case for going the RAM disk direction anymore though. 2.5" solid state drives are slightly cheaper than spinning disks now so there's no reason *not* to use one on a new lane (cheaper not meaning GB-for-GB identical, but for a lane drive the space difference between a 120GB drive and a 250 or 300GB drive is irrelevant). Even for an existing lane the upgrade cost is < $50.

**Comment by flathat** [comment]
>There could also be hardware specific I/O issues where this approach is faster with SSD but slower with spinny disk.

What I was thinking about was evening that out.  To make behaviour between disk types more consistent.  It's the variation in behaviour between lanes that is hard to support/troubleshoot/fix.

**Comment by gohanman** [comment]
I'm skeptical that would really help. It's an extra complication in initial setup, the details of which (and possibly performance characteristics) are going to vary across operating systems. I think it's virtually certain that a non-technical user isn't going to have much luck trying to debug whether the RAM disk is mounted correctly. I'm not sure CORE could realistically detect that kind of problem across OSes either.

I think this would actually be **adding** a layer of software to support/troubleshoot/fix. If SSDs didn't exist or were absurdly expensive then I could see doing a pros/cons look at it, but given the price tag of a solution that's both fast and adds zero complexity it doesn't make much sense to me to compromise. Individuals can do whatever they want but I'd be hesitant to recommend a RAM disk as the ideal setup. I certainly wouldn't want to outright require it and create extra work for those with fast disks.

(I also think we're getting fairly far afield speculating on the runtime characteristics of code that's seen ~5 minutes of testing on just 1 set of hardware; there's a fair amount of testing left to determine whether this is a better solution in any environment before trying to extrapolate to all environments)

**Comment by gohanman** [comment]
I have one production deployment of this now with messages numbered sequentially so POS reads them in the correct order (aaab5866efd7affdbe880cb0696236b4b8755b8a). The driver aggressively checks to see there are no pending messages and the counter can be cleared to avoid wrap-around issues at the max value of a `ushort`. The highest I was able to get that counter with a backlog of messages was 3 so checking could probably be dialed back a bit if needed (say every 10th or even every 50th message).

Subjectively the scale response feels faster. I was not able produce any "beep-of-death" problem by simultaneously scanning & taking items on/off the scale attempting to collide scale status & good beep messages. With such an intermittent error I don't see to declare it fixed w/o a week or even month of testing.

**Comment by joelbrock** [comment]
i'm cautiously optimistic!!
Keep us posted on this, i'm eager to test this out elsewhere.  But i'll  let you guinea pig it for the rest of us a little while longer.

**Comment by gohanman** [comment]
So the `S10` command does in fact work to reboot the scale. The timing is off currently so it requires a subsequent `WAKEUP` to get the POS reading again (the pause between sending `S10` and `S14` has to be long enough to let the scale fully reset).

I haven't been able to reproduce the beep-of-death to verify whether this will resurrect it, but if it can I think that's minor progress. Making the problem go away entirely would be best, but being to reset it with a keypress rather than messing around with power cables is still a step forward.

**Comment by gohanman** [comment]
The random rebooting continues - although it's still apparently a non-fatal issue and the scale keeps working following reboot. I've got a build with both the `S10` and `S14` commands that are trigger by POS sending a message to the driver commented out. If that helps, either the periodic `S14`s triggered periodically by POS are the problem OR there's a flaw in my logic that lets the `S10` command get triggered by accident.

**Comment by gohanman** [comment]
I made a 2nd deployment on a non-SSD machine. There's no observable performance difference that I can notice. Sending constant `S14` status checks still produces a [subjectively] more responsive on screen display. If I had to guess *why* there's no difference, my leading theories would be A) when repeatedly reading/writing the same directory the OS and/or hard drive is smart enough to utilize the drive's cache (and really tiny files will fit easily in available cache) or B) 9600 baud is so absurdly slow by modern computing standards that even a 5400RPM disk has no problem keeping up.

Another deployment bit kind of related to #670. I've been doing partial updates without apparent problems. For example when changing `SPH_Magellan_Scale` and recompiling the driver, the only binary that actually changes is `SPH.dll`. I was able to copy just that single file out to a lane rather than the entire driver.

**Comment by gohanman** [comment]
Another update: I'm now running the updated `SPH_Magellan_Scale` (constant polling w/ `S14`) on the majority of lanes. I'd still call it subjectively more responsive in weighing than `SPH_Magellan_Classic`. This is consistent across a mix of Apache+PHP versions, hard drive types, CPU speeds, RAM quantities, etc.

I still have occasional issues that require power cycling a scale so I've had to re-enable the `WAKEUP` command but I think these are largely environment/hardware problems and not driver/software problems. Issues are most frequent with a scale that's wobbly because one or more of the corners isn't in contact with the physical counter housing OR scales that are stable but significantly not level. What these two scenarios have in common is the scale has trouble getting or holding zero. Remote displays have been extremely valuable in debugging and IMO should be recommended equipment. Barcodes Inc has them for ~$100 which is minor relative to the cost of the scale itself. I think it's a silly place to scrimp. 

I've noticed with scales that aren't well positioned in the counter that they sometimes need to be power cycled repeatedly. Invariably if the scale boots up and the remote display still shows `-  .  lb` it didn't find a proper zero and will continue to have problems. If it reboots and the remote display shows `0.00lb` then it's ready to use. This is all prior to CORE interacting with the scale which points to a non-software issue. I know there's a setting that will have the scale try to re-zero automatically if it's reading under-zero for a certain number of seconds but that seems like more of a band-aid solution compared to getting the scale seated properly in the counter.
